### ISSUE TITLE
The cert dname extraction logic fails on systems using Java keytool on a non-English language system

### ISSUE THREAD

**Issue by Emilien942702** [dana-at-cp/backdoor-apk#41]
Hi,
Facing issues with backdoor-apk, using Kali Linux 2.0 (lib32z1, lib32ncurses5 and lib32stdc++6 are installed and up to date, so as apktool), when it comes to Signing recompiled APK

```
[*] Recompiling original project with backdoor...done.
[*] Generating RSA key for signing...done.
[*] Signing recompiled APK...done.
[!] Failed to sign recompiled APK
```
Here is the intersting part of the log file
```
I: Using Apktool 2.2.0
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether resources has changed...
I: Building resources...
I: Copying libs... (/lib)
I: Building apk file...
I: Copying unknown files/dir...
dname value: 
jarsigner error: java.lang.RuntimeException: certificate exception: Empty issuer DN not allowed in X509Certificates
Forcing cleanup due to a failure or error state!
```

I can provide you logs & can try out what you want
Thanks a lot

**Comment by Emilien942702** [comment]
Just did `unzip -p myapk.apk META-INF/CERT.RSA |keytool -printcert` on the apk i wanted to use, and as i am running Kali in French it's "Propriétaire : " for "Owner".
Changed the script from this : `dname=`$UNZIP -p $ORIG_APK_FILE $orig_rsa_cert |$KEYTOOL -printcert |grep "Owner:" |sed 's/Owner: //g'``
to this : `dname=`$UNZIP -p $ORIG_APK_FILE $orig_rsa_cert |$KEYTOOL -printcert |grep "Propriétaire :" |sed 's/Propriétaire : //g'``
And now it's working fine :)
Maybe you could integrate an option that regexs the right string (Propriétaire :, Owner:) from the keytool princert command :)

**Comment by dana-at-cp** [comment]
@Emilien942702 Apologies. This is a known issue for systems using a native language other than English. This is a duplicate of issue #17 and I haven't spent time working on a better solution. I'll leave this issue open as a bug and will queue it up to be addressed.

**Comment by dana-at-cp** [comment]
@Emilien942702 Can you try running the script out of the "cert_dname_fix" branch? I made a small change on that branch that I think will solve the problem.

**Comment by Emilien942702** [comment]
@dana-at-cp Just tried, got 
```
dname value: Propriétaire : CN=Flynn Macfarlane, OU=Mobile Development, O=General Adaptive Apps Pty Ltd, L=Brisbane, ST=Queensland, C=AU
erreur keytool : java.io.IOException: Invalid keyword "PROPRIÉTAIRE : CN"
Forcing cleanup due to a failure or error state!
```
Think the extraction is right, expect that you need to filter the first strings :)

**Comment by dana-at-cp** [comment]
@Emilien942702 Thanks for testing. It seems that the "[[:alpha:]]" in the regex doesn't account for different languages. :)

```
dname=`$UNZIP -p $ORIG_APK_FILE $orig_rsa_cert |$KEYTOOL -printcert |head -n 1 |sed 's/^[[:alpha:]]\+: //g'`
``

I'll work out another way to do this via regex.

**Comment by dana-at-cp** [comment]
@Emilien942702 There is a new change on the "cert_dname_fix" branch. Can you try it out when you have a few minutes?

**Comment by Emilien942702** [comment]
@dana-at-cp just tried, signing apk is OK but now i get 
```
[*] Signing recompiled APK...done.
[*] Verifying signed artifacts...done.
[*] Aligning recompiled APK...done.
[!] Failed to align recompiled APK
```
And here is the log : (checked, zipalign is in the right folder and to be sure i cloned the whole folder back again from cert_dname_fix branch and relaunched it,  gave the same results)

> jar signed.
> 
> Warning: 
> No -tsa or -tsacert is provided and this jar is not timestamped. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2044-02-23) or after any future revocation date.
> jar verified.
> 
> Warning: 
> This jar contains entries whose certificate chain is not validated.
> This jar contains signatures that does not include a timestamp. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2044-02-23) or after any future revocation date.
> 
> Re-run with the -verbose and -certs options for more details.
> ./backdoor-apk.sh: ligne 512: third-party/android-sdk-linux/build-tools/23.0.3/zipalign: Aucun fichier ou dossier de ce type
> Forcing cleanup due to a failure or error state!

**Comment by dana-at-cp** [comment]
@Emilien942702 Sounds like you are missing a library that will allow you to run the zipalign 32-bit binary on 64-bit Linux. :)

Try this stackoverflow thread to get an idea of what you might be missing or how to troubleshoot further:

http://stackoverflow.com/questions/2716702/no-such-file-or-directory-error-when-executing-a-binary

**Comment by dana-at-cp** [comment]
@Emilien942702 I think the original issue is resolved with the changes on the "cert_dname_fix" branch.

**Comment by dana-at-cp** [comment]
@Emilien942702 The fix is now on the master branch. Merged #46 earlier today.
